pkg/operators/limiter/limiter.go

```diff
@@ -64,83 +64,88 @@ func (l *limiterOperator) InstanceParams() api.Params {
 }
 
 func (l *limiterOperator) InstantiateDataOperator(gadgetCtx operators.GadgetContext, instanceParamValues api.ParamValues) (operators.DataOperatorInstance, error) {
+	limitsPerDs := make(map[string]int)
+
 	// When getting the GadgetInfo, InstantiateDataOperator is called with an
 	// empty instanceParamValues. In such cases, we don't need to validate the
 	// ParamMaxEntries parameter, but just return an instance if there is, at
 	// least, one array data source; otherwise return nil (disabling the
 	// operator).
-	maxEntries, ok := instanceParamValues[ParamMaxEntries]
-	if !ok {
+	if maxEntries, ok := instanceParamValues[ParamMaxEntries]; ok {
+		var err error
+		limitsPerDs, err = apihelpers.GetIntValuesPerDataSource(maxEntries)
+		if err != nil {
+			return nil, fmt.Errorf("parsing %s (%q): %w", ParamMaxEntries, maxEntries, err)
+		}
+		if len(limitsPerDs) == 0 {
+			return nil, fmt.Errorf("missing value for %s", ParamMaxEntries)
+		}
+	} else {
+		found := false
 		for _, ds := range gadgetCtx.GetDataSources() {
 			if ds.Type() == datasource.TypeArray {
-				return &limiterOperatorInstance{}, nil
+				found = true
+				break
 			}
 		}
-		return nil, nil
-	}
+		if !found {
+			gadgetCtx.Logger().Debugf("limiter: no array data sources found")
+			return nil, nil
+		}
 
-	// If ParamMaxEntries is set, it must be valid.
-	valuesPerDs, err := apihelpers.GetIntValuesPerDataSource(maxEntries)
-	if err != nil {
-		return nil, fmt.Errorf("parsing %s (%q): %w", ParamMaxEntries, maxEntries, err)
+		// Disable the limiter for all data sources by default.
+		limitsPerDs[""] = -1
 	}
-	if len(valuesPerDs) == 0 {
-		return nil, fmt.Errorf("missing value for %s", ParamMaxEntries)
+
+	// Other operators could modify the data sources after this point (e.g., combiner),
+	// so subscribe in the pre-start phase where all them are already registered.
+	return &limiterOperatorInstance{
+		limitsPerDs: limitsPerDs,
+	}, nil
+}
+
+func (l *limiterOperator) Priority() int {
+	return Priority
+}
+
+type limiterOperatorInstance struct {
+	limitsPerDs map[string]int
+}
+
+func (l *limiterOperatorInstance) Name() string {
+	return name
+}
+
+func (l *limiterOperatorInstance) PreStart(gadgetCtx operators.GadgetContext) error {
+	if val, ok := l.limitsPerDs[""]; ok && val == -1 {
+		gadgetCtx.Logger().Debug("limiter: disabled for all data sources")
+		return nil
```

Comment:
If the limiter is disabled, why should it reach this point? Shouldn't the logic above return nil so the instance is not created?

I think InstantiateDataOperator could check if the value is -1, if that's the case, return nil. It was implemented as this on main.

https://github.com/inspektor-gadget/inspektor-gadget/blob/e90d457e4dd6d649f2077dc80b5bd05313bd0f81/pkg/operators/limiter/limiter.go#L98-L101

Reply:
That's the only way I found to cover the three cases we have, but I think we should go with a proper solution outside each operator: https://github.com/inspektor-gadget/inspektor-gadget/pull/4050

Reply:
I updated it now that https://github.com/inspektor-gadget/inspektor-gadget/pull/4050 is merged.